tools/mcp/mcp_server.py

```diff
@@ -80,12 +81,83 @@ def build_strict() -> Dict[str, Any]:
 
 @mcp.tool
 def run_tests() -> Dict[str, Any]:
-    """Run tests against main test project (no build)."""
-    rc, out = _run([
+    """Run tests against main test project."""
+    
+    # Build base command without --no-restore to allow restore + build from clean state
+    cmd = [
         "dotnet", "test", "Philips.CodeAnalysis.Test/Philips.CodeAnalysis.Test.csproj",
-        "--configuration", "Release", "--logger", "trx;LogFileName=test-results.trx", "--no-build"
-    ])
-    return {"status": "success" if rc == 0 else "failure", "return_code": rc, "results": out[-8000:]}
+        "--configuration", "Release",
+        "--logger", "trx;LogFileName=test-results.trx"
+    ]
+    
+    # Check if build artifacts exist - if so, skip both restore and build for speed
+    test_bin = BASE_DIR / "Philips.CodeAnalysis.Test" / "bin" / "Release"
+    if test_bin.exists():
+        cmd.extend(["--no-restore", "--no-build"])
+
+    # Use timeout that works within MCP framework limits
+    # MCP framework appears to timeout around 60s, so use 45s for safety
+    # Tests only take ~40s, but from clean state needs build first
+    timeout = 45
+
+    rc, out = _run(cmd, timeout=timeout)
+    
+    # Parse test results from output
+    test_results = {"passed": 0, "failed": 0, "skipped": 0, "total": 0, "duration": ""}
+    test_summary = ""
+    
+    lines = out.splitlines()
+    for line in lines:
+        # Look for the test results summary line like:
+        # "Passed!  - Failed:     0, Passed:  2015, Skipped:     0, Total:  2015, Duration: 40 s"
+        # "Failed!  - Failed:     5, Passed:  2010, Skipped:     0, Total:  2015, Duration: 40 s"
+        if ("Passed!" in line or "Failed!" in line) and "Total:" in line:
+            test_summary = line.strip()
+            # Parse the numeric values
+            failed_match = re.search(r'Failed:\s*(\d+)', line)
+            passed_match = re.search(r'Passed:\s*(\d+)', line)
+            skipped_match = re.search(r'Skipped:\s*(\d+)', line)
+            total_match = re.search(r'Total:\s*(\d+)', line)
+            duration_match = re.search(r'Duration:\s*([^-]+?)(?:\s*-|$)', line)
```

Comment:
security vulnerability in duration_match regex: potential dos due to backtracking